src/python/pants/backend/python/util_rules/lockfile_metadata.py

```diff
@@ -72,42 +121,34 @@ def from_lockfile(
                 "be decoded. " + error_suffix
             )
 
-        def get_or_raise(key: str) -> Any:
-            try:
-                return metadata[key]
-            except KeyError:
-                raise InvalidLockfileError(
-                    f"Required key `{key}` is not present in metadata header for "
-                    f"{lockfile_description}. {error_suffix}"
-                )
+        concrete_class = _concrete_metadata_classes[metadata["version"]]
 
-        requirements_digest = get_or_raise("requirements_invalidation_digest")
-        if not isinstance(requirements_digest, str):
-            raise InvalidLockfileError(
-                f"Metadata value `requirements_invalidation_digest` in {lockfile_description} must "
-                "be a string. " + error_suffix
-            )
+        return concrete_class._from_json_dict(metadata, lockfile_description, error_suffix)
 
-        try:
-            interpreter_constraints = InterpreterConstraints(
-                get_or_raise("valid_for_interpreter_constraints")
-            )
-        except TypeError:
-            raise InvalidLockfileError(
-                f"Metadata value `valid_for_interpreter_constraints` in {lockfile_description} "
-                "must be a list of valid Python interpreter constraints strings. " + error_suffix
-            )
+    @classmethod
+    def _from_json_dict(
+        cls: Type[_T],
+        json_dict: dict[Any, Any],
+        lockfile_description: str,
+        error_suffix: str,
+    ) -> _T:
+        """Construct a `LockfileMetadata` subclass from the supplied JSON dict.
+
+        *** Not implemented. Subclasses should override. ***
```

Comment:
Hm, imo better is to set `@abstractmethod` and subclass `ABC`? MyPy doesn't like the interaction of ABC and dataclasses - if it complains, use `# type: ignore[error-code]`.

Reply:
It's complaining, and it breaks the type checking in subclasses too. That's a bit too much complaining for me to be comfortable with suppressing.